retrieval/target.go

```diff
@@ -304,16 +271,109 @@ func newHTTPClient(cfg *config.ScrapeConfig) (*http.Client, error) {
 }
 
 func (t *Target) String() string {
-	return t.url.Host
+	return t.host()
+}
+
+func (t *Target) client() (*http.Client, error) {
+	t.RLock()
+	defer t.RUnlock()
+
+	return newHTTPClient(t.scrapeConfig)
+}
+
+func (t *Target) interval() time.Duration {
+	t.RLock()
+	defer t.RUnlock()
+
+	return time.Duration(t.scrapeConfig.ScrapeInterval)
+}
+
+func (t *Target) timeout() time.Duration {
+	t.RLock()
+	defer t.RUnlock()
+
+	return time.Duration(t.scrapeConfig.ScrapeTimeout)
+}
+
+func (t *Target) scheme() string {
+	t.RLock()
+	defer t.RUnlock()
+
+	return string(t.labels[model.SchemeLabel])
+}
+
+func (t *Target) host() string {
+	t.RLock()
+	defer t.RUnlock()
+
+	return string(t.labels[model.AddressLabel])
+}
+
+func (t *Target) path() string {
+	t.RLock()
+	defer t.RUnlock()
+
+	return string(t.labels[model.MetricsPathLabel])
+}
+
+func (t *Target) params() url.Values {
+	t.RLock()
+	defer t.RUnlock()
+
+	params := url.Values{}
+
+	for k, v := range t.scrapeConfig.Params {
+		params[k] = make([]string, len(v))
+		copy(params[k], v)
+	}
+	for k, v := range t.labels {
+		if !strings.HasPrefix(string(k), model.ParamLabelPrefix) {
+			continue
+		}
+		ks := string(k[len(model.ParamLabelPrefix):])
+
+		if len(params[ks]) > 0 {
+			params[ks][0] = string(v)
+		} else {
+			params[ks] = []string{string(v)}
+		}
+	}
+
+	return params
+}
+
+// URL returns a copy of the target's URL.
+func (t *Target) URL() *url.URL {
+	return &url.URL{
+		Scheme:   t.scheme(),
+		Host:     t.host(),
+		Path:     t.path(),
+		RawQuery: t.params().Encode(),
```

Comment:
Similar to above, this could return inconsistent information if the Target is updated while this call is ongoing.